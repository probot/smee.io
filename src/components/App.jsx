import React, { Component } from 'react'
import ListItem from './ListItem.jsx'
import get from 'get-value'
import { AlertIcon, PulseIcon, SearchIcon, PinIcon } from '@primer/octicons-react'
import Blank from './Blank.jsx'

export default class App extends Component {
  constructor (props) {
    super(props)
    this.channel = window.location.pathname.substring(1)
    this.storageLimit = 30

    this.handleClear = this.handleClear.bind(this)

    this.ref = `smee:log:${this.channel}`
    this.pinnedRef = this.ref + ':pinned'
    const ref = localStorage.getItem(this.ref)
    const pinnedRef = localStorage.getItem(this.pinnedRef)

    this.state = {
      log: ref ? JSON.parse(ref) : [],
      pinnedDeliveries: pinnedRef ? JSON.parse(pinnedRef) : [],
      filter: '',
      connection: false
    }

    this.togglePinned = this.togglePinned.bind(this)
    this.isPinned = this.isPinned.bind(this)
  }

  componentDidMount () {
    this.setupEventSource()
  }

  setupEventSource () {
    const url = window.location.pathname
    console.log('Connecting to event source:', url)
    this.events = new window.EventSource(url)
    this.events.onopen = this.onopen.bind(this)
    this.events.onmessage = this.onmessage.bind(this)
    this.events.onerror = this.onerror.bind(this)
  }

  onopen (data) {
    this.setState({
      connection: true
    })
  }

  onerror (err) {
    this.setState({
      connection: false
    })
    switch (this.events.readyState) {
      case window.EventSource.CONNECTING:
        console.log('Reconnecting...', err)
        break
      case window.EventSource.CLOSED:
        console.log('Reinitializing...', err)
        this.setupEventSource()
        break
    }
  }

  onmessage (message) {
    console.log('received message!')
    const json = JSON.parse(message.data)

    // Prevent duplicates in the case of redelivered payloads
    const idProp = 'x-github-delivery'
    if (json[idProp] === undefined || this.state.log.findIndex(l => l[idProp] === json[idProp]) === -1) {
      this.setState({
        log: [json, ...this.state.log]
      }, () => {
        localStorage.setItem(this.ref, JSON.stringify(this.state.log.slice(0, this.storageLimit)))
      })
    }
  }

  handleClear () {
    if (confirm('Are you sure you want to clear the delivery log?')) {
      console.log('Clearing logs')
      const filtered = this.state.log.filter(this.isPinned)
      this.setState({ log: filtered })
      if (filtered.length > 0) {
        localStorage.setItem(this.ref, JSON.stringify(filtered))
      } else {
        localStorage.removeItem(this.ref)
      }
    }
  }

  togglePinned (id) {
    const deliveryId = this.state.pinnedDeliveries.indexOf(id)
    let pinnedDeliveries
    if (deliveryId > -1) {
      pinnedDeliveries = [
        ...this.state.pinnedDeliveries.slice(0, deliveryId),
        ...this.state.pinnedDeliveries.slice(deliveryId + 1)
      ]
    } else {
      pinnedDeliveries = [...this.state.pinnedDeliveries, id]
    }

    this.setState({ pinnedDeliveries })
    localStorage.setItem(this.pinnedRef, JSON.stringify(pinnedDeliveries))
  }

  isPinned (item) {
    const id = item['x-github-delivery'] || item.timestamp
    return this.state.pinnedDeliveries.includes(id)
  }

  render () {
    const { log, filter, pinnedDeliveries } = this.state
    let filtered = log
    if (filter) {
      filtered = log.filter(l => {
        if (filter && filter.includes(':')) {
          let [searchString, value] = filter.split(':')
          if (!searchString.startsWith('body')) searchString = `body.${searchString}`
          console.log(l, searchString, value)
          return get(l, searchString) === value
        }
        return true
      })
    }

    const stateString = this.state.connection ? 'Connected' : 'Not Connected'

    const pinnedLogs = filtered.filter(this.isPinned).map((item, i, arr) => {
      const id = item['x-github-delivery'] || item.timestamp
      return <ListItem key={id} pinned togglePinned={this.togglePinned} item={item} last={i === arr.length - 1} />
    })

    const allLogs = filtered.filter(item => !this.isPinned(item)).map((item, i, arr) => {
      const id = item['x-github-delivery'] || item.timestamp
      return <ListItem key={id} pinned={false} togglePinned={this.togglePinned} item={item} last={i === arr.length - 1} />
    })

    return (
      <main>
        <div className='py-2 bgColor-emphasis fgColor-onEmphasis'>
          <div className='container-md p-responsive d-flex flex-items-center flex-justify-between'>
            <h1 className='f4'>Webhook Deliveries</h1>
            <div className='flex-items-right tooltipped tooltipped-w' aria-label={stateString + ' to event stream'}>
              {this.state.connection
                ? <PulseIcon style={{ fill: '#6cc644' }} />
                : <AlertIcon style={{ fill: 'yellow' }} />}
            </div>
          </div>
        </div>

        {log.length > 0
          ? (
            <div className='container-md py-3 p-responsive'>
              <div className='mb-2'>
                <div className='d-flex flex-items-end mb-2'>
                  <label htmlFor='search' className='d-flex flex-items-center f6 text-gray'><SearchIcon height={12} width={12} className='mr-1' /> Filter by</label>
                  <a className='f6' href='https://github.com/jonschlinkert/get-value' target='_blank' rel='noopener noreferrer'>get-value syntax</a>

                  <button onClick={this.handleClear} className='btn btn-sm btn-danger' style={{ marginLeft: 'auto' }}>Clear deliveries</button>
                </div>
                <input
                  type='text'
                  id='search'
                  placeholder='repository.name:probot'
                  value={filter}
                  onChange={e => this.setState({ filter: e.target.value })}
                  className='input input-lg width-full Box'
                />
              </div>
              {pinnedDeliveries.length > 0 && (
                <>
                  <h6 className='d-flex flex-items-center text-gray mb-1'><PinIcon height={12} width={12} className='mr-1' /> Pinned</h6>
                  <ul className='Box list-style-none pl-0 mb-2'>
                    {pinnedLogs}
                  </ul>
                </>
              )}
              <h6 className='d-flex flex-items-center text-gray mb-1'>All</h6>
              {allLogs.length === 0
                ? <div className='Box p-3 note text-center'>All logs are pinned</div>
                : (
                  <ul className='Box list-style-none pl-0'>
                    {allLogs}
                  </ul>
                  )}
            </div>
            )
          : <Blank />}
      </main>
    )
  }
}
