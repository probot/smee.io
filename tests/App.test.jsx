import React from 'react'
import App from '../src/components/App'
import Blank from '../src/components/Blank'
import { shallow } from 'enzyme'
import issuesOpened from './fixtures/issues.opened.json'
import issuesOpenedTwo from './fixtures/issues.opened2.json'

describe('<App />', () => {
  let localStorage, wrapper, consoleLog

  beforeEach(() => {
    localStorage = {
      getItem: jest.fn(),
      setItem: jest.fn(),
      removeItem: jest.fn()
    }

    const EventSource = jest.fn()
    EventSource.CONNECTING = 0
    EventSource.CLOSED = 2

    Object.defineProperties(window, {
      localStorage: {
        value: localStorage,
        writable: true
      },
      EventSource: {
        value: EventSource,
        writable: true
      },
      location: {
        value: {
          pathname: '/CHANNEL'
        }
      }
    })

    console.log = consoleLog = jest.fn()

    wrapper = shallow(<App />)
  })

  describe('render', () => {
    it('renders the blank page', () => {
      expect(wrapper.containsMatchingElement(<Blank />)).toBeTruthy()
    })

    it('renders a list of logs', () => {
      wrapper.setState({ log: [issuesOpened] })
      expect(wrapper.find('ListItem').exists()).toBeTruthy()
    })

    it('respects the filter', () => {
      wrapper.setState({ log: [issuesOpened, issuesOpenedTwo], filter: 'repository.name:probot' })
      expect(wrapper.find('ListItem').length).toBe(1)
    })

    it('respects the filter if it starts with body', () => {
      wrapper.setState({ log: [issuesOpened, issuesOpenedTwo], filter: 'body.repository.name:probot' })
      expect(wrapper.find('ListItem').length).toBe(1)
    })

    it('only filters correct get-value syntax (with :)', () => {
      wrapper.setState({ log: [issuesOpened, issuesOpenedTwo], filter: 'hello' })
      expect(wrapper.find('ListItem').length).toBe(2)
    })

    it('updates the App state when the filter input changes', () => {
      wrapper.setState({ log: [issuesOpened, issuesOpenedTwo] })
      const input = wrapper.find('input#search')
      input.simulate('change', { target: { value: 'hello' } })
      expect(wrapper.state('filter')).toBe('hello')
    })
  })

  describe('onopen', () => {
    it('sets the connection state to true', () => {
      wrapper.instance().onopen()
      expect(wrapper.state('connection')).toBeTruthy()
    })
  })

  describe('onerror', () => {
    it('sets the connection state to false', () => {
      wrapper.instance().onerror()
      expect(wrapper.state('connection')).toBeFalsy()
    })

    it('logs to the console when the state changes to connecting', () => {
      wrapper.instance().events.readyState = 0 // CONNECTING

      wrapper.instance().onerror('error')
      expect(consoleLog).toHaveBeenCalledWith('Reconnecting...', 'error')
    })

    it('logs to the console when the state changes to closed', () => {
      wrapper.instance().events.readyState = 2 // CLOSED

      wrapper.instance().onerror('error')
      expect(consoleLog.mock.calls[1]).toEqual(['Reinitializing...', 'error'])
    })
  })

  describe('onmessage', () => {
    it('adds the new log to the state and localStorage', () => {
      const item = { 'x-github-delivery': 123 }
      const message = { data: JSON.stringify(item) }
      wrapper.instance().onmessage(message)

      expect(wrapper.state('log')).toEqual([item])
      expect(localStorage.setItem.mock.calls[0][0]).toBe('smee:log:CHANNEL')
      expect(localStorage.setItem.mock.calls[0][1]).toMatchSnapshot()
    })

    it('does not add duplicates to the log array or localStorage', () => {
      const item = { 'x-github-delivery': 123 }
      const message = { data: JSON.stringify(item) }
      wrapper.setState({ log: [item] })
      wrapper.instance().onmessage(message)

      expect(wrapper.state('log').length).toBe(1)
      expect(localStorage.setItem).not.toHaveBeenCalled()
    })

    it('ignores duplicate check when no x-github-delivery header is supplied', () => {
      const item = { body: { value: 'Test body 1' }, timestamp: Date.now() }
      const message = { data: JSON.stringify(item) }
      wrapper.setState({ log: [item] })
      wrapper.instance().onmessage(message)

      expect(wrapper.state('log').length).toBe(2)
      expect(localStorage.setItem).toHaveBeenCalled()
    })
  })

  describe('clear', () => {
    beforeEach(() => {
      window.confirm = jest.fn(() => true)
      const item = { 'x-github-delivery': 123 }
      wrapper.setState({ log: [item] })
    })

    it('clears the log state and localStorage', () => {
      wrapper.instance().handleClear()
      expect(wrapper.state('log')).toEqual([])
      expect(localStorage.removeItem).toHaveBeenCalled()
    })

    it('does not clear pinned deliveries', () => {
      wrapper.instance().togglePinned(123)
      wrapper.instance().handleClear()
      expect(wrapper.state('log')).toMatchSnapshot()
      expect(localStorage.setItem.mock.calls[0][1]).toMatchSnapshot()
    })
  })

  describe('togglePinned', () => {
    it('adds a pinned item to the array', () => {
      wrapper.instance().togglePinned(123)
      expect(wrapper.state('pinnedDeliveries')).toEqual([123])
    })

    it('removes a pinned item from the array', () => {
      wrapper.setState({ pinnedDeliveries: [123] })
      wrapper.instance().togglePinned(123)
      expect(wrapper.state('pinnedDeliveries')).toEqual([])
    })

    it('stores the pinnedDeliveries in localStorage', () => {
      wrapper.instance().togglePinned(123)
      expect(localStorage.setItem.mock.calls[0][0]).toBe('smee:log:CHANNEL:pinned')
      expect(localStorage.setItem.mock.calls[0][1]).toMatchSnapshot()

      wrapper.instance().togglePinned(123)
      expect(localStorage.setItem.mock.calls[1][0]).toBe('smee:log:CHANNEL:pinned')
      expect(localStorage.setItem.mock.calls[1][1]).toMatchSnapshot()
    })
  })
})
