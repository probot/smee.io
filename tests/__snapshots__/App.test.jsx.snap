// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> clear does not clear pinned deliveries 1`] = `
Array [
  Object {
    "x-github-delivery": 123,
  },
]
`;

exports[`<App /> clear does not clear pinned deliveries 2`] = `"[123]"`;

exports[`<App /> onmessage adds the new log to the state and localStorage 1`] = `"[{\\"x-github-delivery\\":123}]"`;

exports[`<App /> togglePinned stores the pinnedDeliveries in localStorage 1`] = `"[123]"`;

exports[`<App /> togglePinned stores the pinnedDeliveries in localStorage 2`] = `"[]"`;
